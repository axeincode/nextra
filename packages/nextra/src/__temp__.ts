/* eslint-disable @typescript-eslint/no-var-requires, @typescript-eslint/ban-ts-comment */

// This file has to be targeted to CJS, to keep `require.resolve` untranspiled.
// Otherwise, the file tracing will not work.

const fs = require('fs')
const path = require('path')

// https://github.com/shuding/nextra/pull/1168#issuecomment-1374960179
// Make sure to include all languages in the bundle when tracing dependencies.
// @ts-ignore
const shikiPath = require.resolve('shiki/package.json')
fs.readdir(path.join(shikiPath, '..', 'languages'), () => null)
