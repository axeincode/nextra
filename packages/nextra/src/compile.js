import { compile } from '@mdx-js/mdx'
import remarkGfm from 'remark-gfm'
import { remarkMdxCodeMeta } from 'remark-mdx-code-meta'

import { remarkStaticImage } from './static-image'

export async function compileMdx(source, mdxOptions = {}, nextraOptions = {}) {
  return String(
    await compile(source, {
      jsx: true,
      providerImportSource: '@mdx-js/react',
      remarkPlugins: [
        ...(mdxOptions.remarkPlugins || []),
        remarkGfm,
        remarkMdxCodeMeta,
        nextraOptions.unstable_staticImage && remarkStaticImage
      ].filter(Boolean)
    })
  )
}
