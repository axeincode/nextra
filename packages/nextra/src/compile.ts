import { createProcessor, ProcessorOptions } from '@mdx-js/mdx'
import { Processor } from '@mdx-js/mdx/lib/core'
import remarkGfm from 'remark-gfm'
import rehypePrettyCode from 'rehype-pretty-code'
import { rehypeMdxTitle } from 'rehype-mdx-title'
import readingTime from 'remark-reading-time'
import {
  remarkStaticImage,
  remarkHeadings,
  structurize,
  parseMeta,
  attachMeta
} from './mdx-plugins'
import { LoaderOptions, PageOpts, ReadingTime } from './types'
import theme from './theme.json'
import { truthy } from './utils'

const createCompiler = (mdxOptions: ProcessorOptions): Processor => {
  const compiler = createProcessor(mdxOptions)
  compiler.data('headingMeta', {
    headings: []
  })
  return compiler
}

const rehypePrettyCodeOptions = {
  theme,
  onVisitLine(node: any) {
    // Prevent lines from collapsing in `display: grid` mode, and
    // allow empty lines to be copy/pasted
    if (node.children.length === 0) {
      node.children = [{ type: 'text', value: ' ' }]
    }
  },
  onVisitHighlightedLine(node: any) {
    node.properties.className.push('highlighted')
  },
  onVisitHighlightedWord(node: any) {
    node.properties.className = ['highlighted']
  }
}

export async function compileMdx(
  source: string,
  loaderOptions: Pick<
    LoaderOptions,
    'staticImage' | 'flexsearch' | 'defaultShowCopyCode' | 'readingTime'
  > & {
    mdxOptions?: LoaderOptions['mdxOptions'] &
      Pick<ProcessorOptions, 'jsx' | 'outputFormat'>
  } = {},
  filePath = ''
) {
  const structurizedData = Object.create(null)

  const mdxOptions = loaderOptions.mdxOptions || {}

  const compiler = createCompiler({
    jsx: mdxOptions.jsx || false,
    outputFormat: mdxOptions.outputFormat || 'function-body',
    providerImportSource: '@mdx-js/react',
    remarkPlugins: [
      ...(mdxOptions.remarkPlugins || []),
      remarkGfm,
      remarkHeadings,
      loaderOptions.staticImage && ([remarkStaticImage, { filePath }] as any),
      loaderOptions.flexsearch &&
        structurize(structurizedData, loaderOptions.flexsearch),
      loaderOptions.readingTime && readingTime
    ].filter(truthy),
    rehypePlugins: [
      ...(mdxOptions.rehypePlugins || []),
      parseMeta,
      [
        rehypePrettyCode,
        { ...rehypePrettyCodeOptions, ...mdxOptions.rehypePrettyCodeOptions }
      ],
      [rehypeMdxTitle, { name: '__nextra_title__' }],
      [attachMeta, { defaultShowCopyCode: loaderOptions.defaultShowCopyCode }]
    ]
  })
  try {
    const vFile = await compiler.process(source)
    const result = String(vFile)
      .replace('export const __nextra_title__', 'const __nextra_title__')
      .replace('export default MDXContent;', '')
    const readingTime = vFile.data.readingTime as ReadingTime | undefined
    return {
      result,
      ...(compiler.data('headingMeta') as Pick<
        PageOpts,
        'headings' | 'hasJsxInH1'
      >),
      ...(readingTime && { readingTime }),
      structurizedData
    }
  } catch (err) {
    console.error(`[nextra] Error compiling ${filePath}.`)
    throw err
  }
}
