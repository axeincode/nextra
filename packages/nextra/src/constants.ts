import path from 'node:path'

export const MARKDOWN_EXTENSION_REGEX = /\.mdx?$/

export const IS_PRODUCTION = process.env.NODE_ENV === 'production'

export const LOCALE_REGEX = /\.([a-z]{2}(-[A-Z]{2})?)$/

export const DEFAULT_LOCALE = 'en-US'

export const DEFAULT_CONFIG = {
  staticImage: true,
  flexsearch: {
    codeblocks: true
  }
}

export const OFFICIAL_THEMES = [
  'nextra-theme-docs',
  'nextra-theme-blog'
] as const

export const META_FILENAME = '_meta.json'

export const CWD = process.cwd()

export const MARKDOWN_EXTENSIONS = ['md', 'mdx'] as const

export const PUBLIC_DIR = path.join(CWD, 'public')

export const CACHE_DIR = path.join(CWD, '.next', 'cache')

export const ASSET_DIR = path.join(CWD, '.next', 'static', 'chunks')

export const EXTERNAL_URL_REGEX = /^https?:\/\//
