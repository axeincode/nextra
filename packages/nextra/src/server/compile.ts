import path from 'node:path'
import type { ProcessorOptions } from '@mdx-js/mdx'
import { createProcessor } from '@mdx-js/mdx'
import type { Processor } from '@mdx-js/mdx/lib/core'
import { rendererRich, transformerTwoslash } from '@shikijs/twoslash'
import { remarkMermaid } from '@theguild/remark-mermaid'
import { remarkNpm2Yarn } from '@theguild/remark-npm2yarn'
import type { Program } from 'estree'
import rehypeKatex from 'rehype-katex'
import rehypePrettyCode from 'rehype-pretty-code'
import rehypeRaw from 'rehype-raw'
import remarkFrontmatter from 'remark-frontmatter'
import remarkGfm from 'remark-gfm'
import remarkMath from 'remark-math'
import remarkReadingTime from 'remark-reading-time'
import remarkSmartypants from 'remark-smartypants'
import type { Pluggable, Plugin } from 'unified'
import type {
  FrontMatter,
  LoaderOptions,
  PageOpts,
  ReadingTime,
  StructurizedData
} from '../types'
import {
  CWD,
  DEFAULT_LOCALE,
  ERROR_ROUTES,
  MARKDOWN_URL_EXTENSION_REGEX
} from './constants.js'
import {
  recmaRewriteFunctionBody,
  recmaRewriteJsx
} from './recma-plugins/index.js'
import {
  DEFAULT_REHYPE_PRETTY_CODE_OPTIONS,
  rehypeAttachCodeMeta,
  rehypeBetterReactMathjax,
  rehypeExtractTocContent,
  rehypeIcon,
  rehypeParseCodeMeta
} from './rehype-plugins/index.js'
import {
  remarkCustomHeadingId,
  remarkHeadings,
  remarkLinkRewrite,
  remarkMdxDisableExplicitJsx,
  remarkMdxFrontMatter,
  remarkMdxTitle,
  remarkRemoveImports,
  remarkStaticImage,
  remarkStructurize
} from './remark-plugins/index.js'
import { logger, truthy } from './utils.js'

const cachedCompilerForFormat: Record<
  Exclude<ProcessorOptions['format'], undefined | null>,
  Processor
> = Object.create(null)

type MdxOptions = LoaderOptions['mdxOptions'] &
  Pick<ProcessorOptions, 'jsx' | 'outputFormat'>

// @ts-expect-error -- Without bind is unable to use `remarkLinkRewrite` with `buildDynamicMDX`
// because we already use `remarkLinkRewrite` function to remove .mdx? extensions
const clonedRemarkLinkRewrite = remarkLinkRewrite.bind(null)

type CompileMdxOptions = Pick<
  LoaderOptions,
  | 'staticImage'
  | 'search'
  | 'defaultShowCopyCode'
  | 'readingTime'
  | 'latex'
  | 'codeHighlight'
> & {
  mdxOptions?: MdxOptions
  route?: string
  locale?: string
  filePath?: string
  useCachedCompiler?: boolean
  isPageImport?: boolean
  isPageMapImport?: boolean
}

export async function compileMdx(
  source: string,
  {
    staticImage,
    search,
    readingTime,
    latex,
    codeHighlight,
    defaultShowCopyCode,
    route = '',
    locale,
    mdxOptions = {},
    filePath = '',
    useCachedCompiler,
    isPageImport = true,
    isPageMapImport
  }: CompileMdxOptions = {}
) {
  const {
    jsx = false,
    format: _format = 'mdx',
    outputFormat = 'function-body',
    remarkPlugins,
    rehypePlugins,
    recmaPlugins,
    rehypePrettyCodeOptions
  }: MdxOptions = mdxOptions

  const format =
    _format === 'detect' ? (filePath.endsWith('.mdx') ? 'mdx' : 'md') : _format

  const fileCompatible = filePath ? { value: source, path: filePath } : source
  if (isPageMapImport) {
    const compiler = createProcessor({
      format,
      remarkPlugins: [
        remarkFrontmatter, // parse and attach yaml node
        remarkMdxFrontMatter
      ]
    })
    const vFile = await compiler.process(fileCompatible)
    const content = vFile.toString()

    const index = content.lastIndexOf('function _createMdxContent(props) {')
    const result = content.slice(0, index)

    return { result } as any
  }

  let searchIndexKey: string | null = null
  if (ERROR_ROUTES.has(route)) {
    /* skip */
  } else if (typeof search === 'object') {
    if (search.indexKey) {
      searchIndexKey = search.indexKey(filePath, route, locale)
      if (searchIndexKey === '') {
        searchIndexKey = locale || DEFAULT_LOCALE
      }
    } else {
      searchIndexKey = locale || DEFAULT_LOCALE
    }
  } else if (search) {
    searchIndexKey = locale || DEFAULT_LOCALE
  }

  // https://github.com/shuding/nextra/issues/1303
  const isFileOutsideCWD =
    !isPageImport && path.relative(CWD, filePath).startsWith('..')

  if (isFileOutsideCWD) {
    throw new Error(
      `Unexpected import of "${filePath}" that is outside of working directory, use symlinks instead`
    )
  }

  const isRemoteContent = outputFormat === 'function-body'

  const compiler =
    !useCachedCompiler || isRemoteContent
      ? createCompiler()
      : (cachedCompilerForFormat[format] ??= createCompiler())
  const processor = compiler()

  try {
    const vFile = await processor.process(fileCompatible)

    const data = vFile.data as {
      readingTime?: ReadingTime
      structurizedData: StructurizedData
      title?: string
      frontMatter: FrontMatter
    } & Pick<PageOpts, 'hasJsxInH1'>

    const { readingTime, structurizedData, title, frontMatter, hasJsxInH1 } =
      data
    // https://github.com/shuding/nextra/issues/1032
    const result = String(vFile).replaceAll('__esModule', '_\\_esModule')

    if (typeof title !== 'string') {
      logger.error('`title` is not defined')
    }
    if (!frontMatter) {
      logger.error('`frontMatter` is not defined')
    }

    if (frontMatter.mdxOptions) {
      throw new Error('`frontMatter.mdxOptions` is no longer supported')
    }

    return {
      result,
      title,
      ...(hasJsxInH1 && { hasJsxInH1 }),
      ...(readingTime && { readingTime }),
      ...(searchIndexKey !== null && { searchIndexKey, structurizedData }),
      frontMatter
    }
  } catch (error) {
    console.error(`[nextra] Error compiling ${filePath}.`)
    throw error
  }

  function createCompiler(): Processor {
    return createProcessor({
      jsx,
      format,
      outputFormat,
      providerImportSource: 'nextra/mdx',
      // Fix TypeError: _jsx is not a function for remote content
      development: process.env.NODE_ENV === 'development',
      remarkPlugins: [
        ...(remarkPlugins || []),
        remarkMermaid, // should be before remarkRemoveImports because contains `import { Mermaid } from ...`
        [
          remarkNpm2Yarn, // should be before remarkRemoveImports because contains `import { Tabs as $Tabs, Tab as $Tab } from ...`
          {
            packageName: 'nextra/components',
            tabNamesProp: 'items',
            storageKey: 'selectedPackageManager'
          }
        ] satisfies Pluggable,
        isRemoteContent && remarkRemoveImports,
        remarkFrontmatter, // parse and attach yaml node
        [remarkMdxFrontMatter] satisfies Pluggable,
        remarkGfm,
        format !== 'md' &&
          ([
            remarkMdxDisableExplicitJsx,
            // Replace the <summary> and <details> with customized components
            { whiteList: ['details', 'summary'] }
          ] satisfies Pluggable),
        remarkCustomHeadingId,
        remarkMdxTitle,
        [remarkHeadings, { isRemoteContent }] satisfies Pluggable,
        // structurize should be before `remarkHeadings` because we attach #id attribute to heading node
        search && ([remarkStructurize, search] satisfies Pluggable),
        staticImage && remarkStaticImage,
        readingTime && remarkReadingTime,
        latex && remarkMath,
        // Remove the markdown file extension from links
        [
          clonedRemarkLinkRewrite,
          {
            pattern: MARKDOWN_URL_EXTENSION_REGEX,
            replace: '',
            excludeExternalLinks: true
          }
        ] satisfies Pluggable,
        remarkSmartypants
      ].filter(truthy),
      rehypePlugins: [
        ...(rehypePlugins || []),
        format === 'md' && [
          // To render <details /> and <summary /> correctly
          rehypeRaw,
          // fix Error: Cannot compile `mdxjsEsm` node for npm2yarn and mermaid
          {
            passThrough: ['mdxjsEsm', 'mdxJsxFlowElement', 'mdxTextExpression']
          }
        ],
        [rehypeParseCodeMeta, { defaultShowCopyCode }],
        // Should be before `rehypePrettyCode`
        latex &&
          (typeof latex === 'object'
            ? latex.renderer === 'mathjax'
              ? [rehypeBetterReactMathjax, latex.options, isRemoteContent]
              : [rehypeKatex, latex.options]
            : rehypeKatex),
        ...(codeHighlight === false
          ? []
          : [
              [
                rehypePrettyCode,
                {
                  ...DEFAULT_REHYPE_PRETTY_CODE_OPTIONS,
                  // TODO: For some reason I get Error: Cannot find module 'path' in remote content,
                  // disable twoslash temporarily
                  transformers: isRemoteContent
                    ? []
                    : [
                        transformerTwoslash({
                          renderer: rendererRich(),
                          explicitTrigger: true
                        })
                      ],
                  ...rehypePrettyCodeOptions
                }
              ] as any,
              !isRemoteContent && rehypeIcon,
              rehypeAttachCodeMeta
            ]),
        [rehypeExtractTocContent, { isRemoteContent }]
      ].filter(truthy),
      recmaPlugins: [
        (() => (ast: Program, file) => {
          const mdxContentIndex = ast.body.findIndex(node => {
            if (node.type === 'ExportDefaultDeclaration') {
              return (node.declaration as any).id.name === 'MDXContent'
            }
            if (node.type === 'FunctionDeclaration') {
              return node.id.name === 'MDXContent'
            }
          })

          // Remove `MDXContent` since we use custom HOC_MDXContent
          let [mdxContent] = ast.body.splice(mdxContentIndex, 1) as any

          // In MDX3 MDXContent is directly exported as export default when `outputFormat: 'program'` is specified
          if (mdxContent.type === 'ExportDefaultDeclaration') {
            mdxContent = mdxContent.declaration
          }

          const mdxContentArgument = mdxContent.body.body[0].argument

          file.data.hasMdxLayout =
            !!mdxContentArgument &&
            mdxContentArgument.openingElement.name.name === 'MDXLayout'

          const localExports = new Set(['title', 'frontMatter' /* 'useTOC' */])

          for (const node of ast.body) {
            if (node.type === 'ExportNamedDeclaration') {
              let varName: string
              const { declaration } = node
              if (!declaration) {
                // skip for `export ... from '...'` declaration
                continue
              } else if (declaration.type === 'VariableDeclaration') {
                const [{ id }] = declaration.declarations
                varName = (id as any).name
              } else if (declaration.type === 'FunctionDeclaration') {
                varName = declaration.id.name
              } else {
                throw new Error(`\`${declaration.type}\` unsupported.`)
              }

              if (localExports.has(varName)) {
                Object.assign(node, node.declaration)
              }
            }
          }
        }) satisfies Plugin<[], Program>,
        isRemoteContent ? recmaRewriteFunctionBody : recmaRewriteJsx,
        ...(recmaPlugins || [])
      ].filter(truthy)
    })
  }
}
