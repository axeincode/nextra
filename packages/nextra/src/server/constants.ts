/*
 * Benefit of server/constants - do not include unneeded `path` polyfill in client bundle,
 * while importing constants in client file
 */
import path from 'path'
import type { Property } from 'estree'
import type { NextraConfig } from '../types'

export {
  MARKDOWN_EXTENSION_REGEX,
  ERROR_ROUTES,
  DEFAULT_LOCALE
} from '../constants.js'

export const CWD = process.cwd()

export const PUBLIC_DIR = path.join(CWD, 'public')

export const CHUNKS_DIR = path.join(CWD, '.next', 'static', 'chunks')

export const MARKDOWN_URL_EXTENSION_REGEX = /\.mdx?(?:(?=[#?])|$)/

export const IS_PRODUCTION = process.env.NODE_ENV === 'production'

export const DEFAULT_CONFIG = {
  staticImage: true,
  search: {
    codeblocks: true
  },
  codeHighlight: true,
  autoImportThemeStyle: true
} satisfies Partial<NextraConfig>

export const OFFICIAL_THEMES = ['nextra-theme-docs', 'nextra-theme-blog']

export const META_REGEX = /_meta\.[jt]sx?$/

export const MARKDOWN_EXTENSIONS = ['md', 'mdx'] as const

export const EXTERNAL_URL_REGEX = /^https?:\/\//

export const DEFAULT_LOCALES = ['']

// Experimental, need to deep dive why bundle becomes bigger and there is full
// reload while navigating between pages every time
export const IMPORT_FRONTMATTER = false

export const DEFAULT_PROPERTY_PROPS = {
  type: 'Property',
  kind: 'init',
  method: false,
  shorthand: false,
  computed: false
} satisfies Omit<Property, 'key' | 'value'>

export const TOC_HEADING_REGEX = /^h[2-6]$/
