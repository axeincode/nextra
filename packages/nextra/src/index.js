import { NextraPlugin, pageMapCache } from './plugin'
import { DEFAULT_LOCALE, MARKDOWN_EXTENSION_REGEX } from './constants'

const DEFAULT_EXTENSIONS = ['js', 'jsx', 'ts', 'tsx']
const MARKDOWN_EXTENSIONS = ['md', 'mdx']

const nextra = (...config) =>
  function withNextra(nextConfig = {}) {
    const nextraConfig =
      typeof config[0] === "string"
        ? {
            theme: config[0],
            themeConfig: config[1]
          }
        : config[0]

    const nextraPlugin = new NextraPlugin(nextraConfig)
    const { pageExtensions = DEFAULT_EXTENSIONS } = nextConfig

    if (nextConfig.i18n?.locales) {
      console.log(
        '[nextra] You have Next.js i18n enabled, read here https://nextjs.org/docs/advanced-features/i18n-routing for the docs.'
      )
    }

    // If `i18n.locales` and `i18n.defaultLocale` were not specified,
    // client will receive error - Text content does not match server-rendered HTML.
    // Due to `const { locale } = useRouter()` where `locale` will be `undefined`
    // To fix it we need to explicitly specify `i18n.locales` and `i18n.defaultLocale`
    const i18n = {
      locales: nextConfig.i18n?.locales || [DEFAULT_LOCALE],
      defaultLocale: nextConfig.i18n?.defaultLocale || DEFAULT_LOCALE
    }

    return {
      ...nextConfig,
      i18n,
      pageExtensions: [...pageExtensions, ...MARKDOWN_EXTENSIONS],
      webpack(config, options) {
        config.plugins ||= []
        config.plugins.push(nextraPlugin)

        config.module.rules.push({
          test: MARKDOWN_EXTENSION_REGEX,
          use: [
            options.defaultLoaders.babel,
            {
              loader: 'nextra/loader',
              options: {
                ...nextraConfig,
                locales: i18n.locales,
                defaultLocale: i18n.defaultLocale,
                pageMapCache
              }
            }
          ]
        })

        if (typeof nextConfig.webpack === 'function') {
          return nextConfig.webpack(config, options)
        }

        return config
      }
    }
  }

module.exports = nextra
