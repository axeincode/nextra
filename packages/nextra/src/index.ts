import { NextraPlugin, pageMapCache } from './plugin'
import { DEFAULT_LOCALE, MARKDOWN_EXTENSION_REGEX } from './constants'
import { LoaderOptions, Nextra } from './types'

const DEFAULT_EXTENSIONS = ['js', 'jsx', 'ts', 'tsx'] as const
const MARKDOWN_EXTENSIONS = ['md', 'mdx'] as const

const nextra: Nextra = (...args) =>
  function withNextra(nextConfig = {}) {
    const nextraConfig =
      typeof args[0] === 'string'
        ? {
            theme: args[0],
            themeConfig: args[1] as string
          }
        : args[0]

    const nextraPlugin = new NextraPlugin(nextraConfig)
    const { i18n, pageExtensions = DEFAULT_EXTENSIONS } = nextConfig

    if (i18n?.locales) {
      console.log(
        '[nextra] You have Next.js i18n enabled, read here https://nextjs.org/docs/advanced-features/i18n-routing for the docs.'
      )
    } else if (!i18n?.defaultLocale) {
      // If `i18n.locales` and `i18n.defaultLocale` were not specified,
      // client will receive error - Text content does not match server-rendered HTML.
      // Due to `const { locale } = useRouter()` where `locale` will be `undefined`
      // To fix it we need to explicitly specify `i18n.locales` and `i18n.defaultLocale`
      nextConfig.i18n = {
        ...i18n,
        locales: [DEFAULT_LOCALE],
        defaultLocale: DEFAULT_LOCALE
      }
    }

    return {
      ...nextConfig,
      pageExtensions: [...pageExtensions, ...MARKDOWN_EXTENSIONS],
      webpack(config, options) {
        config.plugins ||= []
        config.plugins.push(nextraPlugin)

        config.module.rules.push({
          test: MARKDOWN_EXTENSION_REGEX,
          use: [
            options.defaultLoaders.babel,
            {
              loader: 'nextra/loader',
              options: <LoaderOptions>{
                ...nextraConfig,
                locales: nextConfig.i18n?.locales,
                defaultLocale: nextConfig.i18n?.defaultLocale,
                pageMapCache
              }
            }
          ]
        })

        if (typeof nextConfig.webpack === 'function') {
          return nextConfig.webpack(config, options)
        }

        return config
      }
    }
  }

module.exports = nextra
