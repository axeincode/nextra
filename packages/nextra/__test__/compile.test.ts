import fs from 'node:fs/promises'
import path from 'node:path'
import { compileMdx } from '../src/compile'
import { it, describe, expect } from 'vitest'
import { CWD } from '../src/constants'

type Name =
  | 'code-h1.mdx'
  | 'code-with-text-h1.mdx'
  | 'dynamic-h1.mdx'
  | 'no-h1.mdx'
  | 'static-h1.mdx'

function loadFixture(name: Name) {
  const filePath = path.join(CWD, '__test__', 'fixture', 'headings', name)
  return fs.readFile(filePath, 'utf8')
}

const mdxOptions = {
  jsx: true,
  outputFormat: 'program' as const
}

describe('process heading', () => {
  it('code-h1', async () => {
    const data = await loadFixture('code-h1.mdx')
    const result = await compileMdx(data, { mdxOptions })
    expect(result).toMatchSnapshot()
  })
  it('code-with-text-h1', async () => {
    const data = await loadFixture('code-with-text-h1.mdx')
    const result = await compileMdx(data, { mdxOptions })
    expect(result).toMatchSnapshot()
  })
  it('static-h1', async () => {
    const data = await loadFixture('static-h1.mdx')
    const result = await compileMdx(data, { mdxOptions })
    expect(result).toMatchSnapshot()
  })
  it('dynamic-h1', async () => {
    const data = await loadFixture('dynamic-h1.mdx')
    const result = await compileMdx(data, { mdxOptions })
    expect(result).toMatchSnapshot()
  })
  it('no-h1', async () => {
    const data = await loadFixture('no-h1.mdx')
    const result = await compileMdx(data, { mdxOptions })
    expect(result).toMatchSnapshot()
  })
})
