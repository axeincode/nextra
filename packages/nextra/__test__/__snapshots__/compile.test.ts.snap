// Vitest Snapshot v1

exports[`process heading > code-h1 1`] = `
{
  "headings": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 16,
              "line": 1,
              "offset": 15,
            },
            "start": {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "inlineCode",
          "value": "codegen.yml",
        },
      ],
      "depth": 1,
      "position": {
        "end": {
          "column": 16,
          "line": 1,
          "offset": 15,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "heading",
      "value": "codegen.yml",
    },
  ],
  "result": "/*@jsxRuntime automatic @jsxImportSource react*/
import {useMDXComponents as _provideComponents} from \\"@mdx-js/react\\";
export const titleText = \\"codegen.yml\\";
function _createMdxContent(props) {
  const _components = Object.assign({
    h1: \\"h1\\",
    code: \\"code\\"
  }, _provideComponents(), props.components);
  return <_components.h1><_components.code>{\\"codegen.yml\\"}</_components.code></_components.h1>;
}
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = Object.assign({}, _provideComponents(), props.components);
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
}
export default MDXContent;
",
  "structurizedData": {},
}
`;

exports[`process heading > code-with-text-h1 1`] = `
{
  "headings": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 16,
              "line": 1,
              "offset": 15,
            },
            "start": {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "inlineCode",
          "value": "codegen.yml",
        },
        {
          "position": {
            "end": {
              "column": 21,
              "line": 1,
              "offset": 20,
            },
            "start": {
              "column": 16,
              "line": 1,
              "offset": 15,
            },
          },
          "type": "text",
          "value": " file",
        },
      ],
      "depth": 1,
      "position": {
        "end": {
          "column": 21,
          "line": 1,
          "offset": 20,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "heading",
      "value": "codegen.yml file",
    },
  ],
  "result": "/*@jsxRuntime automatic @jsxImportSource react*/
import {useMDXComponents as _provideComponents} from \\"@mdx-js/react\\";
export const titleText = \\"codegen.yml file\\";
function _createMdxContent(props) {
  const _components = Object.assign({
    h1: \\"h1\\",
    code: \\"code\\"
  }, _provideComponents(), props.components);
  return <_components.h1><_components.code>{\\"codegen.yml\\"}</_components.code>{\\" file\\"}</_components.h1>;
}
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = Object.assign({}, _provideComponents(), props.components);
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
}
export default MDXContent;
",
  "structurizedData": {},
}
`;

exports[`process heading > dynamic-h1 1`] = `
{
  "hasJsxInH1": true,
  "headings": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 22,
              "line": 8,
              "offset": 153,
            },
            "start": {
              "column": 3,
              "line": 8,
              "offset": 134,
            },
          },
          "type": "text",
          "value": "Posts Tagged with “",
        },
        {
          "attributes": [],
          "children": [],
          "data": {
            "_mdxExplicitJsx": true,
          },
          "name": "TagName",
          "position": {
            "end": {
              "column": 32,
              "line": 8,
              "offset": 163,
            },
            "start": {
              "column": 22,
              "line": 8,
              "offset": 153,
            },
          },
          "type": "mdxJsxTextElement",
        },
        {
          "position": {
            "end": {
              "column": 33,
              "line": 8,
              "offset": 164,
            },
            "start": {
              "column": 32,
              "line": 8,
              "offset": 163,
            },
          },
          "type": "text",
          "value": "”",
        },
      ],
      "depth": 1,
      "position": {
        "end": {
          "column": 33,
          "line": 8,
          "offset": 164,
        },
        "start": {
          "column": 1,
          "line": 8,
          "offset": 132,
        },
      },
      "type": "heading",
      "value": "Posts Tagged with “”",
    },
  ],
  "result": "/*@jsxRuntime automatic @jsxImportSource react*/
import {useMDXComponents as _provideComponents} from \\"@mdx-js/react\\";
export const titleText = \\"Posts Tagged with “”\\";
import {useRouter} from 'next/router';
export const TagName = () => {
  const {tag} = useRouter().query;
  return tag || null;
};
function _createMdxContent(props) {
  const _components = Object.assign({
    h1: \\"h1\\"
  }, _provideComponents(), props.components);
  return <_components.h1>{\\"Posts Tagged with “\\"}<TagName />{\\"”\\"}</_components.h1>;
}
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = Object.assign({}, _provideComponents(), props.components);
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
}
export default MDXContent;
",
  "structurizedData": {},
}
`;

exports[`process heading > no-h1 1`] = `
{
  "headings": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 6,
              "line": 1,
              "offset": 5,
            },
            "start": {
              "column": 4,
              "line": 1,
              "offset": 3,
            },
          },
          "type": "text",
          "value": "H2",
        },
      ],
      "depth": 2,
      "position": {
        "end": {
          "column": 6,
          "line": 1,
          "offset": 5,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "heading",
      "value": "H2",
    },
  ],
  "result": "/*@jsxRuntime automatic @jsxImportSource react*/
import {useMDXComponents as _provideComponents} from \\"@mdx-js/react\\";
function _createMdxContent(props) {
  const _components = Object.assign({
    h2: \\"h2\\"
  }, _provideComponents(), props.components);
  return <_components.h2 id=\\"h2\\">{\\"H2\\"}</_components.h2>;
}
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = Object.assign({}, _provideComponents(), props.components);
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
}
export default MDXContent;
",
  "structurizedData": {},
}
`;

exports[`process heading > static-h1 1`] = `
{
  "headings": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 14,
              "line": 1,
              "offset": 13,
            },
            "start": {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "text",
          "value": "Hello World",
        },
      ],
      "depth": 1,
      "position": {
        "end": {
          "column": 14,
          "line": 1,
          "offset": 13,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "heading",
      "value": "Hello World",
    },
  ],
  "result": "/*@jsxRuntime automatic @jsxImportSource react*/
import {useMDXComponents as _provideComponents} from \\"@mdx-js/react\\";
export const titleText = \\"Hello World\\";
function _createMdxContent(props) {
  const _components = Object.assign({
    h1: \\"h1\\"
  }, _provideComponents(), props.components);
  return <_components.h1>{\\"Hello World\\"}</_components.h1>;
}
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = Object.assign({}, _provideComponents(), props.components);
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
}
export default MDXContent;
",
  "structurizedData": {},
}
`;
